packages/material-ui-unstyled/src/FormControlUnstyled/FormControlUnstyled.tsx

```diff
@@ -0,0 +1,234 @@
+ * For instance, only one input can be focused at the same time, the state shouldn't be shared.
+ *
+ * Demos:
+ *
+ * - [Text Fields](https://material-ui.com/components/text-fields/)
+ *
+ * API:
+ *
+ * - [FormControlUnstyled API](https://material-ui.com/api/form-control-unstyled/)
+ */
+const FormControlUnstyled = React.forwardRef(function FormControlUnstyled(
+  props: FormControlUnstyledProps,
+  ref: React.ForwardedRef<any>,
+) {
+  const {
+    defaultValue,
+    children,
+    className,
+    component,
+    components = {},
+    componentsProps = {},
+    disabled = false,
+    error = false,
+    focused: visuallyFocused = false,
+    onChange,
+    required = false,
+    value,
+    ...other
+  } = props;
+
+  const isControlled = value !== undefined;
+  const [filled, setFilled] = React.useState(() => isFilled(value, defaultValue));
+
+  const [focusedState, setFocused] = React.useState(false);
+  if (disabled && focusedState) {
+    setFocused(false);
+  }
+
+  const focused = visuallyFocused !== undefined && !disabled ? visuallyFocused : focusedState;
+
+  const styleProps: FormControlUnstyledStyleProps = {
+    ...props,
+    disabled,
+    error,
+    filled,
+    focused,
+    required,
+  };
+
+  let registerEffect = () => {};
+  if (process.env.NODE_ENV !== 'production') {
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    const registeredInput = React.useRef(false);
+    registerEffect = () => {
+      if (registeredInput.current) {
+        console.error(
+          [
+            'Material-UI: There are multiple `Input` components inside a FormControl.',
+            'This creates visual inconsistencies, only use one `Input`.',
+          ].join('\n'),
+        );
+      }
+
+      registeredInput.current = true;
+      return () => {
+        registeredInput.current = false;
+      };
+    };
+  }
+
+  React.useEffect(() => {
+    if (isControlled) {
+      setFilled(isFilled(value));
+    }
+  }, [value, isControlled]);
+
+  const handleChange = (event: React.ChangeEvent<NativeFormControlElement>) => {
+    if (!isControlled) {
+      setFilled(isFilled(event.currentTarget.value));
+    }
+
+    onChange?.(event);
+  };
+
+  const childContext: FormControlUnstyledState = {
+    defaultValue,
+    disabled,
+    error,
+    filled,
+    focused,
+    onBlur: () => {
+      setFocused(false);
+    },
+    onChange: handleChange,
+    onFocus: () => {
+      setFocused(true);
+    },
+    registerEffect,
+    required,
+    value,
+  };
+
+  const Root = component ?? components.Root ?? 'div';
+  const rootProps = appendStyleProps(Root, { ...other, ...componentsProps.root }, styleProps);
+
+  return (
+    <FormControlUnstyledContext.Provider value={childContext}>
+      <Root
+        ref={ref}
+        {...rootProps}
+        className={clsx(
+          classes.root,
+          className,
+          rootProps?.className,
```

Comment:
nit: do we need to handle `className` and `rootProps.className` separately? I think `rootProps.className` is sufficient if we stop destructuring `className`.